Use NPM + Bunde modules using Webpack / require and avoid using global/window scope (simplify Makefile)

tchakabam opened the pull request:
Hey

This makes this component an actual NPM package which make it easy to integrate into existing web projects.

First-off this introduces the usage of a require function to actually nest inter-module dependencies and not have to ship everything in window i.e global scope.

It also uses Webpack (its kind off a more advanced Browserify - but there are opinions here) to bundle all your modules into a build file.

It also exposes all the necessary entry points in the index.js files so that downstream applications can use this directly via some kind of require-mechanism and compile/flatten deps themselves.

There is also a static test server `node-static` and a `eslint` integration (which is preferred to jshint) which was used before.

All of this is integrated with the Makefile as a top-level build-tool entry point.

I tried `make demo`, `make dist` and `make tests` from a clean build (except the emscripten stuff) and it works fine.

The Makefile now only drives the Emscripten stuff really, and takes care of some testing, but here I'd really advise for using a tool like Karma that will make this stuff much simpler then what we do in Make now (it also has a driver for Qunit): https://github.com/karma-runner/karma-qunit

tchakabam commented:
TODO/FIXME: 

- [ ] Emscripten build scripts need some renaming in Makefile due to path changes

- [ ] Use Karma-Qunit to drive tests

- [ ] Automate version replacement (and fix current ugliness around demo.js)

tchakabam commented:
To just see the new folder structure go here: https://github.com/tchakabam/ogv.js

tchakabam commented:
Note: You need to have `npm` on your system (it always comes with Node runtime when you install that) but everything is still wrapped into the Makefile, one never needs to call upon it directly.

brion commented:
Awesome! Will check it out in a bit...

spadgos commented on Makefile:
```diff
@@ -3,18 +3,29 @@ BUILDDATE:=$(shell date -u "+%Y%m%d%H%M%S")
+
+JS_FILES := $(shell find src/js -type f -name "*.js")
+
+.FAKE : all clean cleanswf swf js demo democlean tests dist lint
```
Shouldn't this be `.PHONY` ? https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html

tchakabam replied:
i've seen `.fake` or `.FAKE` in other Makefiles for the equal purpose

Mikushi commented:
One question: why? 

There was no issue with ogv.js and you've added build dependencies that I for one don't want to deal with. I mean it's nice and all, but I fail to see the purpose of this.

tchakabam commented:
Hey :) Unfortunately I need to answer your question by disagreeing with some of your statements at first. Yes there were issues from my perspective. These were a) shipping everything in global scope b) having to use the C-preprocessor to build a JavaScript project. This code is very complex and it will make it much simpler to modularize using the paradigm that the `require` function proposes. The only dependency this adds is `npm`, which is a de-facto standard tool in JavaScript development. I don't see how it is hard to "mess" with, but I am happy to help you with anything that might come your way and annoy you with this. Can you explain?

tchakabam commented:
Also add to that potential benefits powerful testing frameworks that can be integrated using node-based tools, or development servers that will allow you to work on a testbench or example without having to call `make` again :) 

I think this project mixes development environments people coming from C/C++ know, and as a web project with a huge JavaScript codebase it definitely needs to get up-to-pace with what is the web and JS in 2015. Of course this worked before, but it was a hack. If you want JS dev taskforce to get onto this, it will be the first step to at least make this use modern paradigms like require. By-the-way I think nested dependencies is something you'd do in any other (script) language as well after all.

I know both environments and worlds (C/C++ vs JS) quite well, and I understand you don't to have to mess with more tools, but trust me this will ease a lot the further developments. I think its already in a great state but there will be more things to be done, especially on the JS part and that has be modern methods used.

And again, you said it has no issues, are you sure? ;) https://github.com/brion/ogv.js/issues

brion commented:
Couple quick notes:

* cortado.jar should maybe be under demo/ rather than assets/ as it's only used in the demo
* the WebGL shaders have been copied directly into WebGLFrameSink.js -- I'd much rather have a build step do the packing into strings so the source files can still be edited :)
* 'const' may be unavailable in IE10 per http://caniuse.com/#feat=const -- I'll have to test this .... oh nevermind those got taken back out I think :D
* will need to double-check that the public class names are still intact in the default usage for back-compat (I can probably drop them for a later 2.0 though)

brion commented:
There's also something wrong in the order of build steps; 'make clean && make' does not build the actual demuxer or decoder modules, but does try to copy them out of the 'build' dir into the dist tree (which of course fails).

brion commented:
Looks like 20f3d00 removed the demuxer/decoder modules as Makefile build dependencies on ogv.js, but did not add them to the webpack configuration, so nothing builds them.

tchakabam commented:
* Emscripten module builds: oops. Fixed that by simply adding these targets back as deps to the build/ogv.js target (webpack actually does not mess with or trigger these builds at all since its done by these shell-scripts)

* Shaders: Modified your file2def tool, it now produces a commonjs module definition :) That is I more or less reinstated the old way to go, only that the output now generates JS (commonjs module that exports a string) and this is then require'd by the respective webgl code. 

* You may want to have a look at the top of `demo.js` to see how I handled backward compatibility i.e still using window namespace if necessary. Everything you need should be there, its just a matter of defining it in the respective scope with the expected name.

* Right now when you use the bundle webpack produces (`build/ogv.js`) the index module (which pojnts to src/ogv.js) is exported to `window.ogv` or whatever global namespace you have. But applications can also require `index.js` themselves and use the ogv module how they prefer. That main module contains all the sub-modules you'd need in an app.

micros123 commented:
Hey @tchakabam, I'm very happy you took it upon you to normalize this library for JS integration via NPM, which is indeed the defacto standard for web development by now. I was looking at the project and it didn't look anything like the way most JS libraries on Github are structured. So I'll just integrate your for directly.
I hope @brion you'll consider merging this pull-request to facilitate web developers to use your great library.

monofonik commented:
Nice work on this @tchakabam, although the build fails for me with a missing JS file - `src/js/WebGLDeepFrameSink.js`?

brion commented:
Ok lemme get this working... re testing now against current :)

brion commented:
refactor-webpack branch has a bunch of build fixes & some tweaks to the loader to be slightly less breaky; I'll refactor it a bit further and then merge it shortly.

I went ahead and re-exposed OGVPlayer, OGVLoader, OGVCompat and OGVVersion (I might be able to re-drop OGVVersion, but I think I need it as a cache-breaker still from the client in MediaWiki's TimedMediaHandler) but fixed the old accidental global exposure of some of the worker proxy classes and stuff.

Will clean the api up further for 2.0 but I don't want to break (semver! :D)

Gotta fix up a couple glitches in the tests, then make sure nothing's surprising going on with the loading/URLs, and I think it should be ready to merge quite soon. :)

brion commented:
(I also had to remove the ref to WebGLDeepFrameSink, curious to see what it is but it looks like it didn't make it in via a git add ;))

brion commented:
Ok seems to be building and running cleanly. :D Thanks again @tchakabam! This is definitely a better way to combine modules, and much less hacky for starters.

I'll be merging more patches & doing more build process refactoring later this weekend & next week, should get a 1.1.0 out shortly and package it up for npm & bower.

Will refactor the API further for a 2.0 in the following weeks, along with splitting out the helper classes (AudioFeeder is already in npm/bower and will get some refactors I'm not done with yet; StreamFile and FrameSink will get same treatment.)

tchakabam commented:
Wow this one took a while but eventually it made it :D thanks i am happy it proves useful to you guys! 

the DeepFrameSink thing was an experimental thing I was doing to be able to do frame-by-frame seeking.

tchakabam commented:
🍻
